### PR TITLE
feature/client-typescript

### PR DESCRIPTION
**Pull request by MasterKale**
This PR includes a **full TypeScript rewrite** of the `@webauthn/client` library. The compiled version of the library still supports use in JavaScript projects, but I've also added the emitting of several **.d.ts** files that support Intellisense and autocompletion of all of the library's methods in both JavaScript- and TypeScript-based projects.

Webpack is still used to bundle everything into a single **main.js** as before. However, Babel has been completely replaced with **ts-loader** for simplified TypeScript configuration while editing and when compiling this library.

Lastly I've also updated the README (in the **client/** folder specifically) to talk about the contents of this library specifically.

### CODE REVIEW

**Comment by MasterKale** [comment]
This PR was partially inspired by https://github.com/wallix/webauthn/issues/16.

I really like the direction this library is going and figured I'd try and contribute some improvements. I wasn't able to really gauge the level of interest in adding more TypeScript support (aside from that PR) so I apologize in advance if I'm overstepping by proposing such a drastic rewrite of half of this library 😅

**Comment by P4sca1** [comment]
Great work! Would be great to also do this for `@webauthn/server`.

File: packages/client/package-lock.json [hunk 1]
```diff
@@ -0,0 +1,3779 @@
+{
```
Comment (eheikes) [inline packages/client/package-lock.json]: This file isn't necessary -- this project uses Lerna and Yarn workspaces.

File: packages/client/README.md [hunk 2]
```diff
@@ -1,73 +1,29 @@
...
 
-Implementation of strong authentication with the webauthn standard and FIDO2.
-Strong authentication is an authentication method using a physical key.
+This library contains convenience methods intended to simplify interactions with a browser's Webauthn API.
 
-For a more thorough introduction see these two nice articles:
-
-- [introduction](https://medium.com/@herrjemand/introduction-to-webauthn-api-5fd1fb46c285)
-- [verifying fido2 responses](https://medium.com/@herrjemand/verifying-fido2-responses-4691288c8770)
+This front end library compliments this repo's [@webauthn/server](../server/) NodeJS library.
```
Comment (eheikes) [inline packages/client/README.md]: nitpick 
```suggestion
This front end library complements this repo's [@webauthn/server](../server/) NodeJS library.
```

**Review by eheikes — COMMENTED** [review]
Nice work @MasterKale. I noticed a couple minor things but otherwise looks good.

File: packages/client/package-lock.json (hunk 1 above)
Comment (MasterKale) [reply packages/client/package-lock.json]: Apologies, I've never used Lerna before! I need to quickly read up on it and Yarn's workspaces to get a better feel for it.

After some reading I'll aim to remove this file, restore the old Yarn lock file, and then update the Yarn lock file with these new+updated dependencies 😄

File: packages/client/README.md (hunk 2 above)
Comment (MasterKale) [reply packages/client/README.md]: Aha, nice catch! I'll fix this as well.

**Comment by MasterKale** [comment]
@P4sca1 That's my hope! I decided to tackle the client codebase first as it was a much lower LOE to convert to TS in a backwards-compatible way. If this PR gets in then some of the interfaces included will simplify porting the server codebase over as well :v:

**Comment by MasterKale** [comment]
I'm trying to get up to speed with Yarn workspaces and Lerna to make sure my work doesn't break the publishing workflow (it's my first time with either). However, I can't make heads or tails of the fact that, while workspaces seem to keep a single repo-wide **yarn.lock** file, this project also has yarn.lock files in both the **client/** and **server/** packages.

It seems what I need to do is delete **client/yarn.lock**, then include in this PR changes to the root **yarn.lock** that occur when I run `yarn install`. If not, what's the proper way to update the lock file to track my changes to **client**'s dependencies?

**Comment by P4sca1** [comment]
Afaik there shouldn’t be a yarn.lock in client and server packages. There should only be one in the root directory.

**Comment by eheikes** [comment]
> Afaik there shouldn’t be a yarn.lock in client and server packages. There should only be one in the root directory.

I think that's legacy cruft that was never cleaned up in #17.

> It seems what I need to do is delete client/yarn.lock, then include in this PR changes to the root yarn.lock that occur when I run yarn install.

You could also run `lerna bootstrap`, which delegates to `yarn`.

**Comment by MasterKale** [comment]
> Afaik there shouldn’t be a yarn.lock in client and server packages. There should only be one in the root directory.

Great, thanks for clearing that up! I'll go ahead and clean up those two yarn.lock files, and then update the root one.

I've been pretty busy lately but I haven't forgotten about this! I do plan on getting these fixes in the next day or two :)

**Comment by MasterKale** [comment]
Alright, sorry about the delay! I believe I've addressed all of the outstanding issues we talked about above. This includes removing both **client/yarn.lock** and **server/yarn.lock** directories, and an updated **yarn.lock** in the root of the project.

Barring any other suggested changes, this is ready to go!

**Comment by MasterKale** [comment]
I'm bumping this so it doesn't get too stale...

**Comment by P4sca1** [comment]
@MasterKale Have you found another package for Webauthn or why have you closed this pr?

**Comment by MasterKale** [comment]
@P4sca1 I closed this PR because I'm releasing my own pair of front-end and back-end WebAuthn libraries under the **"WebAuthntine"** name:

https://github.com/MasterKale/WebAuthntine

I took some inspiration on the project's structure from this repo, but the code is entirely based on a side project I built last year while diving deep into the API to learn how it works. And best of all everything is 100% TypeScript!

**Comment by P4sca1** [comment]
This is great! I always thought about writing my own code for webauthn in typescript, but haven't found the time. Very excited about your project. I hope it gets more attention than this project.

**Pull request closed without merging**